File: scripts/audit-wiki-data.mjs
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const dataDir = path.join(root, 'docs', 'data');
const batchDir = path.join(dataDir, 'skill-catalog-batches');
const failures = [];
const warnings = [];

function readJson(file) {
  try {
    return JSON.parse(fs.readFileSync(file, 'utf8'));
  } catch (error) {
    failures.push(`${path.relative(root, file)}: invalid JSON (${error.message})`);
    return null;
  }
}

const canonicalFile = path.join(dataDir, 'skills.json');
const verifiedFile = path.join(dataDir, 'verified-skills.json');
const canonical = readJson(canonicalFile);
const verified = readJson(verifiedFile);

if (!canonical) failures.push('Canonical skills database is unavailable.');
if (!verified) failures.push('Verified skills database is unavailable.');

const records = [];
for (const record of canonical?.records ?? []) records.push({ ...record, origin: 'canonical' });
for (const record of verified?.records ?? []) records.push({ ...record, origin: 'verified' });

if (fs.existsSync(batchDir)) {
  for (const filename of fs.readdirSync(batchDir).filter(name => name.endsWith('.json')).sort()) {
    const file = path.join(batchDir, filename);
    const batch = readJson(file);
    if (!batch) continue;
    if (!batch.category || !batch.source || !Array.isArray(batch.records)) {
      failures.push(`${path.relative(root, file)}: batch requires category, source and records[]`);
      continue;
    }
    if (Number.isInteger(batch.count) && batch.count !== batch.records.length) {
      failures.push(`${path.relative(root, file)}: count=${batch.count} but records=${batch.records.length}`);
    }
    if (!batch.source.startsWith('http')) warnings.push(`${path.relative(root, file)}: source is not an HTTP URL`);
    for (const name of batch.records) records.push({ name, class: batch.category, origin: filename });
  }
}

const seen = new Map();
for (const record of records) {
  const key = `${record.name}|${record.class ?? ''}|${record.subcategory ?? ''}`.toLowerCase();
  if (seen.has(key)) {
    const previous = seen.get(key);
    warnings.push(`duplicate candidate: ${record.name} [${record.origin}] also appears in ${previous}`);
  } else {
    seen.set(key, record.origin);
  }
  if (!record.name) failures.push(`${record.origin}: record has no name`);
}

if (canonical?.category_counts) {
  for (const [category, expected] of Object.entries(canonical.category_counts)) {
    const matching = records.filter(record => record.origin === `${category}.json` || record.class === category || record.subcategory === category);
    if (expected < 0) failures.push(`${category}: negative expected count`);
    if (!matching.length) warnings.push(`${category}: no directly attributable loaded batch records; category may rely on canonical seed data`);
  }
}

console.log(`Skill/data audit: ${failures.length} failure(s), ${warnings.length} warning(s)`);
for (const message of failures) console.error(`FAIL: ${message}`);
for (const message of warnings.slice(0, 100)) console.warn(`WARN: ${message}`);
if (warnings.length > 100) console.warn(`WARN: ${warnings.length - 100} additional warnings omitted`);
process.exitCode = failures.length ? 1 : 0;

File: scripts/audit_wiki_data.py
#!/usr/bin/env python3
"""Audit structured Dragon Ball Xenoverse 2 wiki data.

The audit is intentionally conservative: it reports duplicate identities, missing
required fields, empty provenance, invalid verification states, category-count
mismatches, and obvious cross-domain references. It never promotes data to
verified and never mutates source records.
"""
from __future__ import annotations

import json
from collections import Counter, defaultdict
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[1]
DATA = ROOT / "docs" / "data"

REQUIRED_BY_FILE = {
    "skills.json": ["name", "class", "subcategory", "verification_status", "sources"],
    "verified-skills.json": ["name", "verification_status", "sources"],
    "mentors-record-layer.json": ["id", "name", "verification_status", "sources"],
    "parallel-quests-record-layer.json": ["id", "number", "name", "verification_status", "sources"],
}
VALID_STATUS = {"indexed", "partially_verified", "verified"}


def load_json(path: Path) -> Any:
    with path.open(encoding="utf-8") as fh:
        return json.load(fh)


def records_from(payload: Any) -> list[dict[str, Any]]:
    if isinstance(payload, dict) and isinstance(payload.get("records"), list):
        return [r for r in payload["records"] if isinstance(r, dict)]
    if isinstance(payload, list):
        return [r for r in payload if isinstance(r, dict)]
    return []


def effective_value(record: dict[str, Any], field: str, payload: Any) -> Any:
    """Resolve a record value, allowing supported file-level defaults."""
    if field in record:
        return record[field]
    if field == "verification_status" and isinstance(payload, dict):
        return payload.get("verification_status")
    return None


def main() -> int:
    findings: list[dict[str, Any]] = []
    summary = Counter()
    scanned = 0

    for filename, required in REQUIRED_BY_FILE.items():
        path = DATA / filename
        if not path.exists():
            findings.append({"severity": "error", "type": "missing_file", "file": filename})
            summary["errors"] += 1
            continue
        try:
            payload = load_json(path)
        except json.JSONDecodeError as exc:
            findings.append({"severity": "error", "type": "invalid_json", "file": filename, "message": str(exc)})
            summary["errors"] += 1
            continue

        records = records_from(payload)
        scanned += len(records)
        ids: dict[str, list[int]] = defaultdict(list)
        names: dict[str, list[int]] = defaultdict(list)
        for idx, record in enumerate(records):
            record_id = record.get("id")
            name = record.get("name")
            if record_id:
                ids[str(record_id)].append(idx)
            if name:
                names[str(name).strip().casefold()].append(idx)

            missing = [
                field
                for field in required
                if effective_value(record, field, payload) is None
            ]
            if missing:
                findings.append({"severity": "error", "type": "missing_required_fields", "file": filename, "record": name or record_id, "fields": missing})
                summary["errors"] += 1

            sources = effective_value(record, "sources", payload)
            if "sources" in record and not sources:
                findings.append({"severity": "error", "type": "empty_sources", "file": filename, "record": name or record_id})
                summary["errors"] += 1

            status = effective_value(record, "verification_status", payload)
            if status is not None and status not in VALID_STATUS:
                findings.append({"severity": "error", "type": "invalid_verification_status", "file": filename, "record": name or record_id, "status": status})
                summary["errors"] += 1

        for value, positions in ids.items():
            if len(positions) > 1:
                findings.append({"severity": "error", "type": "duplicate_id", "file": filename, "id": value, "records": positions})
                summary["errors"] += 1
        for value, positions in names.items():
            if len(positions) > 1:
                findings.append({"severity": "warning", "type": "duplicate_name", "file": filename, "name": value, "records": positions})
                summary["warnings"] += 1

        if isinstance(payload, dict) and isinstance(payload.get("category_counts"), dict):
            actual = Counter(str(r.get("subcategory") or r.get("class") or "") for r in records)
            for category, declared in payload["category_counts"].items():
                count = actual.get(category, 0)
                if count != declared:
                    findings.append({"severity": "warning", "type": "category_count_mismatch", "file": filename, "category": category, "declared": declared, "actual": count})
                    summary["warnings"] += 1

    report = {
        "schema_version": "1.1",
        "generated_by": "scripts/audit_wiki_data.py",
        "game": "Dragon Ball Xenoverse 2",
        "records_scanned": scanned,
        "summary": {
            "errors": summary["errors"],
            "warnings": summary["warnings"],
            "findings": len(findings),
        },
        "policy": {
            "errors_block_promotion": True,
            "warnings_require_review": True,
            "no_automatic_verification": True,
            "no_count_inflation_from_partial_sources": True,
            "file_level_verification_defaults_supported": True,
        },
        "findings": findings,
    }
    out = DATA / "audit-latest.json"
    out.write_text(json.dumps(report, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    print(json.dumps(report, indent=2, ensure_ascii=False))
    return 1 if summary["errors"] else 0


if __name__ == "__main__":
    raise SystemExit(main())
